### ISSUE TITLE
How to run the front end UI?

### ISSUE THREAD

**Issue by scheung38** [rhgraysonii/ocr_tutorial#4]
WIthin DigitalOcean new droplet, as per DOCKER_INSTALL.md

Ran

```
boot2docker init
boot2docker: command not found
```

and

```
(env)root@sebtest:~/python_ocr_tutorial# docker build --rm -t flask-docker-ocr .
The following packages have unmet dependencies:
 libtool : Depends: cpp but it is not going to be installed
           Depends: libc6-dev but it is not going to be installed or
                    libc-dev
E: Unable to correct problems, you have held broken packages.
2015/09/30 11:36:43 The command [/bin/sh -c apt-get install -y autoconf automake libtool] returned a non-zero code: 100
```

Not sure how DOCKER.INSTALL.md is related as it is for OSX when it is Ubuntu in Digital Ocean? Can I not install in my Mac OSX?

**Comment by ybur-yug** [comment]
Hey @scheung38 , `boot2docker` is an OSX-only utility. What it's purpose is in this context is to get a verrry small linux image mounted inside OSX in order to be able to run docker. 

To set up Docker on Ubuntu, I would follow [this guide](https://docs.docker.com/installation/ubuntulinux/).

However, do note that Docker has had major changes since this was written a year ago. Now you will need to use `docker-machine` and `docker-compose` over `docker` and `fig`. So, feel free to ask more questions in here and I'll help as I can. I have both an ubuntu machine and a mac so I should be able to help you out. I'm going to close the issue since its unrelated to the repository really, but feel free to keep commenting. Just tag me so I get a notification.

**Comment by scheung38** [comment]
Yes I did see now it is docker-machine instead, so the installation of

```
boot2docker init
boot2docker start
```

is no longer valid and deprecated?

Thanks I was a bit  confused how to launch the app and see the front end UI so I don't have to do it via curl. Secondly how to config Digital Ocean Droplet(or any remote machines) to display as for example

```
gedit ocr.py

(gedit:30536): Gtk-WARNING **: cannot open display:
```

or can I ask to update to show how to launch front end UI as make a Heroku deployment so I can play with the front end, as I already tried setenv DISPLAY:0.0 and still showing the above. 

Also how to push images from Mac to Digital Ocean?

Ok I see, so boot2docker tool for Mac OSX, but Digital Ocean is only Ubuntu right? 

Was a bit confused that after running _app.sh another subdirectory was created with similar name to top level python_ocr_tutorial, named  python_ocr_tutorial/ocr-tutorial-0? 

```
~/python_ocr_tutorial/python_ocr_tutorial-0/flask_server
~/python_ocr_tutorial/flask_server# 

~/python_ocr_tutorial/flask_server# ls -al
total 52
drwxr-xr-x 4 root root 4096 Sep 30 12:00 .
drwxr-xr-x 8 root root 4096 Sep 30 11:47 ..
-rw-r--r-- 1 root root 1344 Sep 30 10:41 app.py
-rw-r--r-- 1 root root  986 Sep 30 10:41 cli.py
-rw-r--r-- 1 root root  106 Sep 30 10:41 config.py
-rw-r--r-- 1 root root   64 Sep 30 12:00 error.log
-rw-r--r-- 1 root root 1004 Sep 30 10:41 fabfile.py
-rw-r--r-- 1 root root    0 Sep 30 10:41 __init__.py
-rw-r--r-- 1 root root  335 Sep 30 10:41 ocr.py
-rw-r--r-- 1 root root  807 Sep 30 12:00 ocr.pyc
-rw-r--r-- 1 root root   19 Sep 30 10:41 Procfile
-rw-r--r-- 1 root root   85 Sep 30 10:41 Procfile.dev
drwxr-xr-x 7 root root 4096 Sep 30 10:41 static
drwxr-xr-x 2 root root 4096 Sep 30 10:41 templates

~/python_ocr_tutorial/python_ocr_tutorial-0/flask_server# ls -altotal 32
drwxrwxr-x 3 root root 4096 Jan  6  2015 .
drwxrwxr-x 3 root root 4096 Jan  6  2015 ..
-rw-rw-r-- 1 root root  727 Jan  6  2015 app.py
-rw-rw-r-- 1 root root  106 Jan  6  2015 config.py
-rw-rw-r-- 1 root root 1004 Jan  6  2015 fabfile.py
-rw-rw-r-- 1 root root    0 Jan  6  2015 __init__.py
-rw-rw-r-- 1 root root   19 Jan  6  2015 Procfile
-rw-rw-r-- 1 root root   85 Jan  6  2015 Procfile.dev
drwxrwxr-x 7 root root 4096 Jan  6  2015 static
```

I have same setup, Mac locally at home and Digital Ocean Ubuntu 14 droplet created.

**Comment by ybur-yug** [comment]
`gedit` is a graphical utility that requires a GUI to utilize. If you want to edit text on a linux machine the most common pre-installed editors are `nano` and `vi`. So, `nano ocr.py` would allow you to edit it.

With regard to getting a frontend running, you may be best off simply writing a small JS snipped that hits the API.

It looks like the tools are somewhat deprecated with all the Docker changes, so how about this --
I will test out getting all the things running on a personal DO machine, and document the process. I can then update the repository with this info so that it is up-to-date and you can follow easily since it will be the same OS/setup.

I will get back to you once I do this, probably later today.

**Comment by ybur-yug** [comment]
I reopened it since the core is on me (lack of documentation)

**Comment by scheung38** [comment]
Thanks gedit was just an example, because same if trying to launch any apps such as google chrome in remote machine would be the same issue as well

```
Gtk-WARNING **: cannot open display:
```

Did try 

```
setenv DISPLAY:0.0
```

but also didn't work.

But the templates and static folders are already there? Not sure why suggesting to do front end? 

So it is either:
Docker Machine for running the `docker-machine` binary
Docker Engine for running the `docker` binary
Docker Compose for running the `docker-compose` binary
?

Thanks for your prompt response and help.

**Comment by ybur-yug** [comment]
I would just abandon docker altogether, this can be done in a simpler manner if you just wanna get up and running. To get the application running on a digital ocean box (just the API) on a clean machine logged in as root run this script:

``` bash
apt-get update
apt-get install git\
              python-pip
sudo apt-get build-dep python-imaging
git clone https://github.com/ybur-yug/python_ocr_tutorial.git
cd python_ocr_tutorial
pip install -r requirements.txt
python flask_server/app.py
```

This will get you all you need to run. Now,  here is what I would do to hit it simply. If you have a local file, `index.html` and embed this javascript:

``` javascript
    $.ajax({
      type: "POST",
      url: "you_ip_address:5000/v1/ocr",
      contentType: "application/json",
      dataType: "json",
      data : JSON.stringify({ "image_url" : some_image_url }),
      success: function(result) {
        console.log(result);
      },
```

and replace `some_image_url` with an image url, and `your_ip_address1` with the IP of a digital ocean machine you have the server running on, this AJAX request will return the results and log it to the js console. If you would like, you dont even have to have it embedded in an `index.html` file, you could do it all from the console in firefox/chrome.

Does this help?

**Comment by scheung38** [comment]
How to provide the actual image from local Mac to the app, which is running locally inside Digital Ocean environment? Or am I mistaken? But you are referring to a chrome instance inside Digital Ocean and not from my Mac right? If that's not the case not sure if there is any connection as both are local. But if it is then not sure how to launch chrome inside Digital Ocean environment. 

```
url: "you_ip_address" 
```

is that my Mac IP or Digital Ocean IP address?

**Comment by ybur-yug** [comment]
The way the server works is it downloads the image from a URL manually. So sending files locally is not supported. What you must provide instead is an image url to an image elsewhere. Imgur meme's have a lot of text that is easily ocr'd for testing, or you can also you the hard-link to the image from the repository used for testing on github.

the url part is your digital ocean machine that is running the flask server's ip. Not your local machine.
What I am referring to specifically is firing up the API on a digital ocean instance, then hitting it from its public url, which is its ip address followed by the port, ie `193.42.52.3:5000`. So if you run this on the DO machine, and hit the API manually using that JS snippet from chrome on your own computer, you would be good to go.

**Comment by ybur-yug** [comment]
This is also assuming you have tesseract, leptonica, etc installed outside of docker. which you can do easily with this: https://github.com/ybur-yug/python_ocr_tutorial/blob/137c35719a2d82c625d53cbfc687648315cd6cd5/_run.sh

**Comment by scheung38** [comment]
But running the _app.py now generates another flask_server subdirectory as mentioned above? now getting two app.py files, one in each of the following:

~/python_ocr_tutorial/python_ocr_tutorial-0/flask_server
~/python_ocr_tutorial/flask_server

**Comment by ybur-yug** [comment]
Figured I'd write the process out to hit it clearly since its scattered across the comments. 

## Running The API and Hitting It Without Docker
1. SSH into a digital ocean machine
2. Run this script:

``` bash
sudo apt-get update
sudo apt-get install git\
              python-pip
sudo apt-get build-dep python-imaging
git clone https://github.com/ybur-yug/python_ocr_tutorial.git
cd python_ocr_tutorial
pip install -r requirements.txt
```

And then this:

```
# Install Dependencies
sudo apt-get update
sudo apt-get install -y autoconf automake libtool
sudo apt-get install -y libpng12-dev
sudo apt-get install -y libjpeg62-dev
sudo apt-get install -y g++
sudo apt-get install -y libtiff4-dev
sudo apt-get install -y libopencv-dev libtesseract-dev git cmake build-essential libleptonica-dev
sudo apt-get install -y liblog4cplus-dev libcurl3-dev
sudo apt-get install -y python2.7-dev tk8.5 tcl8.5 tk8.5-dev tcl8.5-dev
sudo apt-get build-dep -y python-imaging --fix-missing
sudo apt-get install -y imagemagick


# Building Leptonica
wget http://www.leptonica.org/source/leptonica-1.70.tar.gz
tar -zxvf leptonica-1.70.tar.gz
cd leptonica-1.70/
./autobuild
./configure
make
sudo make install
sudo ldconfig

# Building Tesseract
cd ..
wget https://tesseract-ocr.googlecode.com/files/tesseract-ocr-3.02.02.tar.gz
tar -zxvf tesseract-ocr-3.02.02.tar.gz
cd tesseract-ocr/
./autogen.sh
./configure
make
sudo make install
sudo ldconfig

# Set Environment Variable
TESSDATA_PREFIX=/usr/local/share/

# Download the relevant Tesseract English Language Packages
cd ..
wget https://tesseract-ocr.googlecode.com/files/tesseract-ocr-3.02.eng.tar.gz
tar -xf tesseract-ocr-3.02.eng.tar.gz
sudo cp -r tesseract-ocr/tessdata $TESSDATA_PREFIX
```

Now that all dependencies are handled, we can run the server:

``` bash
python flask_server/app.py
```

And the server will be running on `DIGITAL_OCEAN_MACHINES_IP:5000`

And to hit it we can open up the developer js console in chrome or firefox, and type in:

``` python
    $.ajax({
      type: "POST",
      url: "DIGITAL_OCEAN_MACHINE_IP:5000/v1/ocr",
      contentType: "application/json",
      dataType: "json",
      data : JSON.stringify({ "image_url" : IMAGE_URL_FROM_INTERNET }),
      success: function(result) {
        console.log(result);
      },
```

Where DIGITAL_OCEAN_MACHINE_IP is your box that is running the server's IP and IMAGE_URL_FROM_INTERNET is some picture with text from imgur, the projects github repo, whatever. Any valid image url works. If the browser will load it you should be good.

From there, you will receive the string back that will log what was parsed in the server to the js console.

**Comment by ybur-yug** [comment]
I added sudos to two apt-get commands in an edit but its good now. Confirmed to work on a clean DO box i just spun up. This is basically the annotated bash history of getting it working.

**Comment by scheung38** [comment]
Which two apt-get commands? Could I also install node inside Digital Ocean and within node, launch this script since I cannot get chrome to launch in Digital Ocean

**Comment by ybur-yug** [comment]
apt-get update and apt-get install pip/vim, the first 2

Well you can hit the IP address of the digital ocean box from any device with an internet connection. I did it myself on my macbook pro using osx's chome. The API is running and exposed to the public. The IP address acts like the url to the api and you just specify the port (in this case, 5000) and you can hit it anywhere. You could even fire up IRB and just do

``` ruby
require 'net/http'

Net::HTTP.post_form(URI('http://DIGITAL_OCEAN_IP:5000/v1/ocr'), { image_url: "YOUR_IMAGE_URL" })
#=> and you have the parsed response
```

**Comment by ybur-yug** [comment]
Since its on an API and live you can just hit it with whatever client you choose, anything that can POST an HTTP request is able to fine. No authentication or anything is set up.

**Comment by scheung38** [comment]
Sorry IRB as in, ok Interactive Ruby Shell...can you run ruby in Pycharm? If not, possible to get Ruby plugins? Can only ruby via command lines in PyCharm.

```
require 'net/http'

Net::HTTP.post_form(URI('http://DIGITAL_OCEAN_IP:5000/v1/ocr'), { image_url: "YOUR_IMAGE_URL" })
```

if `YOUR_IMAGE_URL` is in my desktop then it would be?

$HOME/myimg.jpg?
$HOME/myimg.png?

**Comment by ybur-yug** [comment]
No, it will NOT download an image from your desktop, as it has no means to know how to connect to it. You would have to build an HTML form that submits the request to the API and change the way the server itself works to do this.

What you need to do is take an image like this

https://realpython.com/images/blog_images/ocr/ocr.jpg

As in, that URL

and then feed it in using whatever your method of choice to make an HTTP POST request is, with the parameters "{ 'image_url': 'https://realpython.com/images/blog_images/ocr/ocr.jpg' }"
and it would work if you are POSTing to the endpoint that is `
`DIGITAL_OCEAN_MACHINE_IP:5000/v1/ocr`

**Comment by scheung38** [comment]
Could we not use the front end that is already there in templates folder?

**Comment by ybur-yug** [comment]
Yes, you could even just run it locally and specify the right URL. I thought you hadn't done that for a specific reason since its already provided and can be run without the server actually being on your machine.

**Comment by scheung38** [comment]
Would it be easier to build everything locally in my Mac? Just not sure how to install all the dependencies since it was designed for Linux. Could I request build installation but for Mac.

**Comment by ybur-yug** [comment]
to run it on OSX you only need to install boot2docker, then run

```
$ boot2docker init
$ boot2docker start
$ docker build --rm -t flask-docker-ocr .
$ docker run -p 80:5000 flask-docker-ocr
```

You shouldnt need docker-machine or anything else (I think, let me know)

**Comment by ybur-yug** [comment]
and it will have all the things working local (frontend and server both)

**Comment by scheung38** [comment]
Reason is not knowing how to install the dependencies in Mac OSX. So apt-get would be replaced with brew?

**Comment by scheung38** [comment]
Any others?

**Comment by ybur-yug** [comment]
docker will take care of everything, you dont need to manually install anything except boot2docker and docker itself.

**Comment by scheung38** [comment]
Boot2docker is deprecated for Mac OSX too

I see the conversion from

$boot2docker init is now -> docker-machine create? 
$boot2docker start -> docker-machine start ?
$docker build --rm -t flask-docker-ocr .
$ docker run -p 80:5000 flask-docker-ocr

Have not got any to work yet?
after $docker build --rm -t flask-docker-ocr . and after a long build, says successfully built but none of the dependencies are present? Was expecting to see all the dependencies to be in the same folder structure as flask_server, not sure where they went?

Back in DO

```
~/python_ocr_tutorial/flask_server# python app.py 
 * Running on http://0.0.0.0:5000/
127.0.0.1 - - [30/Sep/2015 20:01:55] "POST /v1/ocr HTTP/1.1" 200 -
```

also in DO
~/python_ocr_tutorial/flask_server# curl -X POST http://localhost:5000/v1/ocr -d '{"image_url": "some_url"}' -H "Content-Type: application/json"
{
"error": "only .jpg files, please"
}
still in DO
~/python_ocr_tutorial/flask_server# curl -X POST http://localhost:5000/v1/ocr -d '{"image_rl": "https://www.realpython.com/images/blog_images/ocr/ocr.jpg"}' -H "Content-Type: application/json"

{
  "output": "ABCDE\nFGH I J\nKLMNO\nPQRST"
}

But in Mac IRB:

$ irb
Ignoring gem-wrappers-1.2.7 because its extensions are not built.  Try: gem pristine gem-wrappers --version 1.2.7
1.9.3-p194 :001 > require 'net/http'
 => true 
1.9.3-p194 :002 > Net::HTTP.post_form(URI('http://178.62.18.164:5000/v1/ocr'), { image_url: 'https://www.realpython.com/images/blog_images/ocr/ocr.jpg'})
 => #<Net::HTTPOK 200 OK readbody=true>

curl -X POST http://178.62.18.164:5000/v1/ocr -d '{'image_url': 
'https://www.realpython.com/images/blog_images/ocr/ocr.jpg'}' -H 'Content-Type: application/json'

in summary no success yet building in OSX

Could you make a Heroku deployment to help?

![screen shot 2015-09-30 at 23 17 34](https://cloud.githubusercontent.com/assets/1273014/10208250/091b6daa-67ca-11e5-99a5-21cf6045e2b2.png)

**Comment by ybur-yug** [comment]
The error you are getting on the server means that you are not sending a URL that ends in `.jpg`, which is required for the server. It looks like you are up and running on DO and got the actual output.
